lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -2233,90 +2253,60 @@ namespace {
         return false;
       }
 
+      case ActorIsolationRestriction::CrossDistributedActorSelf:
       case ActorIsolationRestriction::CrossActorSelf: {
         // If a cross-actor reference is to an isolated actor, it's not
         // crossing actors.
-        if (getIsolatedActor(base))
+        auto isolatedActor = getIsolatedActor(base);
+        if (isolatedActor)
           return false;
 
-        return diagnoseNonSendableTypesInReference(
-            memberRef, getDeclContext()->getParentModule(), memberLoc,
-            ConcurrentReferenceKind::CrossActor);
-      }
-
-      case ActorIsolationRestriction::DistributedActorSelf: {
-        // distributed actor isolation is more strict;
-        // we do not allow any property access, or synchronous access at all.
-        // FIXME: We can collapse a much of this with the ActorSelf case.
-        bool continueToCheckingLocalIsolation = false;
-        // Must reference distributed actor-isolated state on 'self'.
-        //
-        // FIXME(78484431): For now, be loose about access to "self" in actor
-        // initializers/deinitializers for distributed actors.
-        // We'll want to tighten this up once we decide exactly
-        // how the model should go.
-        auto isolatedActor = getIsolatedActor(base);
-        if (!isolatedActor &&
+        bool performDistributedChecks =
+            isolation.getActorType()->isDistributedActor() &&
             !(isolatedActor.isActorSelf() &&
               member->isInstanceMember() &&
-              isActorInitOrDeInitContext(getDeclContext()))) {
-          // cross actor invocation is only ok for a distributed or static func
+              isActorInitOrDeInitContext(getDeclContext())
+            ) &&
+            !member->getAttrs().hasAttribute<NonisolatedAttr>() &&
+            !member->getAttrs().hasAttribute<DistributedActorIndependentAttr>();
+
+        if (performDistributedChecks) {
           if (auto func = dyn_cast<FuncDecl>(member)) {
             if (func->isStatic()) {
-              // FIXME: rather, never end up as distributed actor self isolated
-              //        at all for static funcs
-              continueToCheckingLocalIsolation = true;
+              // no additional checks for static functions
             } else if (func->isDistributed()) {
-              tryMarkImplicitlyAsync(
-                  memberLoc, memberRef, context,
-                  ImplicitActorHopTarget::forInstanceSelf());
-              tryMarkImplicitlyThrows(memberLoc, memberRef, context);
-
-              // distributed func reference, that passes all checks, great!
-              continueToCheckingLocalIsolation = true;
+              tryMarkImplicitlyAsync(memberLoc, memberRef, context,
```

Comment:
I think our refactoring goal here should be to unify `tryMarkImplicitlyAsync` and `tryMarkImplicitlyThrows`, and give it enough information to determine when to add the qualifiers. For example, if `isPotentiallyIsolated` is true for a function on a distributed actor, we don't need the function to be `distributed` and should not mark it `throws`, but we might need to mark it `async`.

Reply:
Yeah that sounds right... they are applied in similar situations, just need to feed it enough flags. 

Noted!